# Confusion Matrix

Imagine Building a model for binary judgment, Is this mail spam or not? ,  Should we hire this candiate or not?

All the result lies in one of the four categories namely:- 

- **True Positive :-**  " This message is spam and we predicted it as spam"
- **False Positive (Type 1 Error):-** " This message is not spam and we predicted it as spam"
- ** False Negative (Type 2 Error):-** " This message is spam and we predicted it as not spam"
- **True Negative :- ** "This message is not spam and we predicted it as not spam"

**We often represent these as counts in confusion matrix :-**

|                  | **Spam**       | **Not Spam** |
|:----------------:|:--------------:|:------------:|
|Predict "spam"    | True Positive  |False Positive|
|Predict "Not Spam"| False Negative |True Negative |

In [1]:
# We use these in calculating various Statistics about model per formances

tp , fp , fn , tn = 70 , 4930, 13930, 981070

def accuracy (tp , fp , fn , tn):
    correct = tp + tn
    total = tp + fp + fn + tn
    return correct/total

print(accuracy(tp , fp , fn , tn))



0.98114


98.114% Seems like a pretty impressive number. But clearly this is not a good test, which means that we probably shouldn't put a lot of credence in raw accuracy.

**It's common to look at the combination of *Pricision* and *recall* **

- Precision measures how accurate our positive predictions were:

$Precision = \frac{tp}{tp + fp}$

- And recall measures what fraction of the positives our model idetified:-

$Recall = \frac{tp}{tp + fn}$

In [2]:
def precision (tp , fp , fn , tn):
    return tp / (tp + fp)

print(precision(tp, fp, fn, tn))

def recall (tp , fp , fn , tn):
    return tp / (tp + fn)

print(recall(tp , fp , fn, tn))

0.014
0.005


These are both terrible numbers, reflecting  that this is a terrible model.
Sometimes *precision* and *recall* are combined into the *F1* Score, which is defined as:

${F1\_score} = \frac{2.P.R}{P + R}$

In [3]:
def f1_score(tp , fp , fn , tn):
    p = precision(tp , fp , fn , tn)
    r = recall(tp , fp , fn , tn)
    return 2*p*r/(p + r)

print(f1_score(tp , fp , fn , tn))

0.00736842105263158


This is the **Harmonic mean** of precision and recall and necessarily lies between them (precision and recall)